retry.go

```diff
@@ -89,3 +98,84 @@ func ExponentialBackoff(backoff Backoff, condition ConditionFunc) error {
 		return NewTimeoutError(errors.New("Timed out while retrying"))
 	}
 }
+
+type auroraThriftCall func() (resp *aurora.Response, err error)
+
+// Duplicates the functionality of ExponentialBackoff but is specifically targeted towards ThriftCalls.
+func (r *realisClient) ThriftCallWithRetries(thriftCall auroraThriftCall) (*aurora.Response, error) {
+	var resp *aurora.Response
+	var clientErr error
+
+	backoff := r.config.backoff
+	duration := backoff.Duration
+
+	for i := 0; i < backoff.Steps; i++ {
+
+		// If this isn't our first try, backoff before the next try.
+		if i != 0 {
+			adjusted := duration
+			if backoff.Jitter > 0.0 {
+				adjusted = Jitter(duration, backoff.Jitter)
+			}
+
+			r.logger.Printf("An error occurred during thrift call, backing off for %v before retrying\n", adjusted)
+
+			time.Sleep(adjusted)
+			duration = time.Duration(float64(duration) * backoff.Factor)
+		}
+
+		// Only allow one go-routine make use or modify the thrift client connection.
+		r.lock.Lock()
+		resp, clientErr = thriftCall()
```

Comment:
Is it guaranteed this call will never panic?

if it does it will leave the lock unavailable for anyone else, in addition to all other damage.

Reply:
@kkrishna has raised this point as well before. As it exists, we can't add defer because the function gets called in a loop. I'll wrap these calls in anonymous function to create a new stack so that defer gets called right after the call is made. Tested this theory here: https://play.golang.org/p/GcAJJeHfS95

Reply:
👍